### DISCUSSION TITLE
Kuma UI v2 RFC

### DISCUSSION THREAD

**Discussion by poteboy** [kuma-ui/kuma-ui#409] [Announcements]
## Introduction

Kuma UI v2 will prioritize **enhancing flexibility and compatibility with the ecosystem**, rather than focusing solely on performance. This version is designed to support a variety of coding styles while ensuring compatibility with the latest technologies like RSC.

## Current State of Kuma UI v1

[`css`](https://www.kuma-ui.com/docs/API/css) and [`styled`](https://www.kuma-ui.com/docs/API/styled) API currently only support zero-runtime approach. These APIs require a compiler to function, and they do not allow variable interpolation. 

```tsx
// Current use without interpolation support
const staticStyle = css`
  color: red;
  font-size: 16px;
`;

// Attempting interpolation (not supported)
const color = 'blue';
const dynamicStyle = css`
  color: ${color}; // This results in an error in v1
`;
```

Headless components such as [`<Box />`](https://www.kuma-ui.com/docs/Components/Box) utilize a [Hybrid Approach](https://www.kuma-ui.com/docs/Concepts/Hybrid) that allows for variable interpolation. However, being internally processed as Client Components, they are not compatible with Server Components.

```tsx
// Processed as a Server Component
function Static() {
  return (
    <Box p={12} /> 
  );
}

// Processed as a Client Component
const padding = 12
function Dynamic() {
  return (
    <Box p={padding} /> 
  );
}
```

## Strategic Direction for Kuma UI v2

We are designed with the philosophy that **the compiler should only serve as an optimization tool, not as a necessity**. Functions provided by `@kuma-ui/core` will be usable with or without compiler optimization., similar to the [React Compiler's](https://react.dev/blog/2024/02/15/react-labs-what-we-have-been-working-on-february-2024) ideology. 

Due to technical constraints, however, the `css` API requires a compiler and will therefore be moved to `@kuma-ui/macro`. This new package will centralize all compiler-dependent functionalities, ensuring that features offered by `@kuma-ui/core` are optimized for performance but fully functional without compiler intervention.

`@kuma-ui/macro` will handle advanced styling functionalities that require the compiler. All plugins, previously distributed across different packages like `@kuma-ui/vite` and `@kuma-ui/next-plugin`, will now be centralized under `@kuma-ui/macro` for better consistency and maintenance.

In short:

- Plugins that were previously separate for each bundler, like `@kuma-ui/vite` and `@kuma-ui/webpack-plugin`, are now consolidated into `@kuma-ui/macro`.
- Our fundamental policy is to ensure that **all functionalities provided by Kuma UI can operate without a compiler**. However, the `css` API specifically requires compiler intervention, so all compiler-dependent features are aggregated under `@kuma-ui/macro`.
- Features offered by `@kuma-ui/core` are accessible without a compiler, but when a compiler is used, it automatically optimizes the process. 
 
Here, the term "compiler" refers to the transpilation process within the bundle process, similar to how Babel or SWC operates, and by integrating plugins, Kuma intervenes in the compile process.

## Technical Enhancements

We will introduce significant improvements to our APIs to enhance both functionality and developer experience. Key enhancements include:

- **Interpolation Support in `styled` and `css` APIs**: We will enable variable interpolation across both APIs, greatly expanding their flexibility and utility. 

```tsx
// With interpolation support
const color = 'blue';
const dynamicStyle = css`
  color: ${color}; // Now valid in v2
`;
```

- **Compiler-Dependent and Independent Features**: The `css` API, requiring compiler support for advanced features, will be part of `@kuma-ui/macro`. In contrast, the `styled` API and `<Box />` will allow runtime interpolations without compiler dependency, fully compatible with Server Components.

```tsx
// This will work as a Server Components
function Component() {
  const [padding ,setPadding] = useState(0)
  return (
    <Box p={padding} /> 
  );
}
```

- **Integration with React v19 Features**: Embracing React v19, [especially the new ](https://ja.react.dev/reference/react-dom/components/style)[<style />](https://ja.react.dev/reference/react-dom/components/style)[ component](https://ja.react.dev/reference/react-dom/components/style), will enable efficient and native styling within React's framework. This alignment enhances server component compatibility and optimizes performance.

To introduce these new features, **the minimum peer dependency version for React will need to be v19**, as this feature utilizes the `<style />` component introduced in that version. We would appreciate your feedback or opinions on this requirement.

On top of that, to enable interpolation within the css API, several steps are necessary:

1. **Hoisting Dependent Variables**: Variables are hoisted to the top level to ensure they are statically analyzable.
2. **Variable Execution in Node.js VM**: The hoisted variables are executed within a Node.js VM to retrieve their values.
3. **Literal Substitution**: The original variables are replaced with the retrieved literals, ensuring the styles can be compiled and optimized.

We currently use [WyW-in-JS](https://wyw-in-js.dev/) for compiler processes but are developing a Rust-based alternative, [jaku](https://github.com/poteboy/jaku), to address performance and cache issues. Jaku will transform ASTs using [oxc-parser](https://oxc-project.github.io/ja/docs/guide/usage/parser.html) and will be distributed via WASM. We invite developers skilled in Rust, Terser, and Minify to help enhance jaku's development.

## Handling of Headless Components

In v2, handling of headless components like `<Box />` will support three main styling strategies to ensure full compatibility with React Server Components (RSC):

- **Static Styling**: Components with static values continue to use static analysis, bundling CSS as external assets.

```tsx
<Box color="red" /> // CSS bundled externally
// Compiled to as follows:
// <div class="kuma-xxx" />
```

- **Variable Interpolation**: When variables are embedded, components utilize CSS variables and inline styles for compilation, fully compatible with RSC.

```tsx
const color = 'red';
<Box color={color} />
// Compiled to as follows:
// <div class="kuma-xxx" style="--color-red: red;" />
// .kuma-xxx { color: var(--color-red); }
```

- **Dynamic Props via Spread Syntax**: For components using spread syntax, which prevents static analysis, Kuma UI leverages React 19’s <style /> component to maintain RSC compatibility.

```tsx
<Box {...props} />
```

Furthermore, considering that Kuma UI components are primarily simple HTML tag wrappers customizable via themes, the proposal for v2 is to streamline component offerings to just five core components: `<Box />`, `<Stack />`, `<Spacer />`, `<Text />`, and `<Button />`. This change aims to simplify the library while maintaining versatility.

**Community Feedback Request**: We are considering this reduction in component variety to streamline development and usage. We welcome feedback on this proposed change to ensure it meets the needs of our users.

### Optimization via Compiler for Static Asset

The `css` API, when enhanced by the compiler, is the most efficient at runtime due to its ability to manage static assets and perform tree shaking. However, it may increase build times since it involves executing variables within a VM during HMR.

```tsx
const staticStyle = css`
  color: red;
  font-size: 16px;
`;
```

### Performance with and without Interpolation

When no interpolation is used and the compiler is present, the performance of `css`, `styled`, and `Box` remains highly optimized and comparable to fully static solutions.

```tsx
// Static usage with compiler optimization
<Box color="red" />
const Component = styled.div`color: red;`
const className = css`color: red;`
```

When interpolation is involved, particularly in the `styled` API, styles are processed without compiler, leading to lower performance compared to compiler-enhanced styles. This method provides flexibility at the cost of runtime efficiency.

```tsx
// Dynamic variables with interpolation in styled API
const dynamicColor = useState('blue')[0];
const dynamicStyle = styled('div')`
  color: ${dynamicColor};
`;
```

The choice of styling method in v2 depends on the specific needs for performance and flexibility. Compiler enhancements offer the best performance, while runtime processing offers maximum flexibility at a potential cost to performance.

## Conclusion and Community Feedback Request ✅

As we move towards finalizing Kuma UI v2, **we eagerly seek input from our community to enhance and refine our offerings**. Whether you have suggestions for new features, modifications to existing ones, or additional implementation ideas, your feedback is invaluable.

You can also share your thoughts and contributions through our [Discord server](https://discord.gg/QrsQ4EPp7G) or by reaching out on [X](https://twitter.com/kuma__ui).

**Comment by dougg0k** [comment]
In your interp examples, you only show variables directly put into the string, is there going to be a way to access the variable inside if you pass it as a prop?

https://styled-components.com/docs/basics#passed-props

**Comment by leeeunsang-tmobi** [comment]
In the current implementation, CSS extraction from the style prop only works when the component is a Kuma UI component:

https://github.com/kuma-ui/kuma-ui/blob/%40kuma-ui/core%401.6.0/packages/compiler/src/compile.ts#L42

Would it be possible to extract CSS based solely on the presence of the style prop, regardless of whether the component is a Kuma UI component?
I believe this would make Kuma UI’s hybrid approach significantly more versatile—for example, when building component library packages on top of Kuma UI.
